Update: Add exceptionPatterns to id-length rule (fixes #13094)

sodaMelon opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[X] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

[issue #113094](https://github.com/eslint/eslint/issues/13094)
<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

I continued the [PR #13099](https://github.com/eslint/eslint/pull/13099)
>to quote from  #13099
>>I added an option named exceptionPatterns to id-length, which takes as value an array of the allowed patterns.
>>If the option is not defined it takes an empty array as default value.
>>Then I declared a boolean matchesExceptions that gets true if one of the patterns tests the name.
>>And added that boolean to the condition.
---
Based on comments in #13099, I modified some codes about it.
* Unnecessary 'parserOption'  was removed from ```id-length.js(test)```
* Two valid tests were added. They test with multiple exception patterns.
* Two invalid tests were added, too. They test with an identifier that doesn't match configured pattern. 
(One tests ```tooLongError```, another tests ```tooShrotError```. )

* To avoid creating new RegExp in "return value", I exatracted a function.

* ```docs/rules/id-length.md``` was updated about ```exceptionPatterns```
 
#### Is there anything you'd like reviewers to focus on?
I want know how these test-cases are suitable. 
I thought them in my own way.........🤔 but not sure.. 😂

yeonjuan commented:
@sodaMelon Would you mind sign with [CLA](https://github.com/eslint/eslint/pull/13576#issuecomment-674785398) ?

sodaMelon commented:
> @sodaMelon Would you mind sign [CLA](https://github.com/eslint/eslint/pull/13576#issuecomment-674785398) ?

Ok

> @sodaMelon Would you mind sign [CLA](https://github.com/eslint/eslint/pull/13576#issuecomment-674785398) ?

Ok, I did it now.

yeonjuan commented on lib/rules/id-length.js:
```diff
@@ -63,8 +70,21 @@ module.exports = {
+        function matchesExceptions(name) {
+            return exceptionPatterns.some(
+                pattern => new RegExp(pattern, "u").test(name)
+            );
```
I think it would be better to create all RegExp instance in the "create(context)", because it's not changed after generated from the option "exceptionPatterns".

yeonjuan commented on docs/rules/id-length.md:
```diff
@@ -217,6 +218,29 @@ const { x } = foo;
+(x_value) => {return x_value * x_value};
+var [x_value] = arr;
+const { x_value } = foo;
+const { a: x_value } = foo;
```
Can we change examples to allowed cases affected by the option  `exceptionPatterns `?.
`x_value` seems correct example regardless of the option `exceptionPatters` option.

* [online-demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgaWQtbGVuZ3RoOiBbXCJlcnJvclwiXSovXG4vKmVzbGludC1lbnYgZXM2Ki9cbnZhciB4X3ZhbHVlID0gNTtcbmZ1bmN0aW9uIGxvbmdGdW5jdGlvbl9zZW5kKCkgeyByZXR1cm4gNDI7IH1cbm9iai54X2xvbmdOYW1lID0gZG9jdW1lbnQuYm9keTtcbnZhciBmb28gPSBmdW5jdGlvbiAoeF92YWx1ZSkgeyAvKiBkbyBzdHVmZiAqLyB9O1xudHJ5IHtcbiAgICBkYW5nZXJvdXNTdHVmZigpO1xufSBjYXRjaCAoeF92YWx1ZSkge1xuICAgIC8vIGlnbm9yZSBhcyBtYW55IGRvXG59XG4oeF92YWx1ZSkgPT4ge3JldHVybiB4X3ZhbHVlICogeF92YWx1ZX07XG52YXIgW3hfdmFsdWVdID0gYXJyO1xudmFyIHsgeF92YWx1ZSB9ID0gZm9vO1xudmFyIHsgYTogeF92YWx1ZSB9ID0gZm9vOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=) (no error)
```js
/*eslint id-length: ["error"]*/
/*eslint-env es6*/
var x_value = 5;
function longFunction_send() { return 42; }
obj.x_longName = document.body;
var foo = function (x_value) { /* do stuff */ };
try {
    dangerousStuff();
} catch (x_value) {
    // ignore as many do
}
(x_value) => {return x_value * x_value};
var [x_value] = arr;
var { x_value } = foo;
var { a: x_value } = foo;
```

sodaMelon replied:
You are right. 
```longFunction_send``` isn't correct, too.
Because there is no default value for ```max-length``` in id-length rule

Maybe I thought it has a default value like 10.
Sorry for the mistake. 

Now i could think of right cases of the rule. like [a-d] or [ E | S | L | i | n | t]. 
Perhaps it will be simpler.

mdjermanovic commented on docs/rules/id-length.md:
```diff
@@ -82,6 +82,7 @@ This rule has an object option:
 * `"properties": always` (default) enforces identifier length convention for property names
 * `"properties": never` ignores identifier length convention for property names
 * `"exceptions"` allows an array of specified identifier names
+* `"exceptionPatterns"` allows an array of identifiers matching the pattern.
```
```suggestion
* `"exceptionPatterns"` array of strings representing regular expression patterns, allows identifiers that match any of the patterns.
```

mdjermanovic commented on lib/rules/id-length.js:
```diff
@@ -63,8 +70,19 @@ module.exports = {
+         * @returns {boolean} if the string is a match
+         * @private
+         */
+        function matchesExceptions(name) {
```
The name of this function might be confusing since we also have the `exceptions` array, so it isn't clear from its name whether this function checks `exceptions`, or `exceptionPatterns`, or maybe both.

Maybe we could rename this function to `matchesExceptionPattern`?

sodaMelon replied:
You are right. It will be better to avoid confusion using that. Thanks for the advice!

mdjermanovic commented on tests/lib/rules/id-length.js:
```diff
@@ -77,7 +77,11 @@ ruleTester.run("id-length", rule, {
+        { code: "let foo = { [a + b]: 1 };", options: [{ properties: "always" }], parserOptions: { ecmaVersion: 6 } },
+        { code: "function BEFORE_send() {};", options: [{ min: 3, max: 5, exceptionPatterns: ["^BEFORE_"] }] },
+        { code: "function BEFORE_send() {};", options: [{ min: 3, max: 5, exceptionPatterns: ["^BEFORE_", "send$"] }] },
+        { code: "function BEFORE_send() {};", options: [{ min: 3, max: 5, exceptionPatterns: ["^BEFORE_", "^A", "Z*"] }] },
```
It would be nice to also add a valid test with multiple patterns where the first pattern doesn't match.

sodaMelon replied:
Ok, I will do this.

mdjermanovic reviewed (changes requested):
Thanks for working on this! Looks great, I have just a few small suggestions.

sodaMelon commented:
I modified the function name to ```matchesExceptionPattern```
And added a valid test that doesn't match with the first pattern. (that matched with the second.)

While making this test, I thought  "Z*" of the test case from the above test not look good.  So I changed it to "^Z". 
(maybe it clearer!)

mdjermanovic reviewed (approved):
LGTM, thanks!
